### ISSUE TITLE
[4.0.5] [BUG] New CreateOperation and UpdateOperation undocumented breaking change

### ISSUE THREAD

**Issue by drionix** [Laravel-Backpack/CRUD#2104]
# Bug report

### What I did

1. Add extra request data before performing CrudOperations. In this case, I inject 'author_id'  directly through ArticleController before performing CreateOperation (I did not present the 'author_id' field to the user due to security concern).
2. Use MenuCRUD add-ons to create menu item and input the menu link.

### What I expected to happen

1. In previous Backpack version (3.x), I could achieve this by simply adding this piece of code
```
public function store(StoreRequest $request)
{	
	$request->request->add(['author_id'=> backpack_user()->id]);
        return parent::storeCrud($request);
}
```
to the controller.

2. In previous Backpack version (3.x), MenuCRUD save whatever link I input for each menu item.

### What happened

1. In Backpack v4.x, I cannot directly add extra request data through the controller before performing CreateOperation. This is caused by this line of code
`$item = $this->crud->create($this->crud->getStrippedSaveRequest());`
in `CreateOperation@store`. Particularly, because `getStrippedSaveRequest()` only returns fields which are presented to the user through `$this->crud->addFields()`.

2. In Backpack 4.x, since the URL field for each menu item is not a real field (only part of `page_or_link` field type), it cannot be saved into database because Backpack 4.x new CrudOperations only returns 'real' fields. 

To be clear, there are actually 2 pseud-fields within `page_or_link` field, namely the `(menu) type (page or link)` field, and the `page_id/link` field. However, the `getStrippedSaveRequest()` only returns the `(menu) type (page or link)` as 'real' field and dismiss the `page_id/link` field. Therefore, I can't input any link into the menu item in MenuCRUD.

### What I've already tried to fix it
It seems like `getStrippedSaveRequest()` function is not compatible with lots of Backpack 3.x projects.

### Backpack, Laravel, PHP, DB version
Backpack 4.x-dev
Laravel 6
PHP 7.2.4

When I run ```php artisan backpack:version``` the output is:

**Comment by evanrthompson** [comment]
I'm also getting tripped up by getStrippedSaveRequest() and tldr; I think the ability to define a field as optional/sometimes during setupXXXOperation would be useful here.

## What I want to do
We have a very simple user management page with first, last, email, password and password confirmation fields. With backpack V3 I was able to submit that page, see if the password fields contained text, and remove those fields from the request if they hadn't been altered so that the update wouldn't mess with them. 

## How I'm getting boned
Now however, regardless of what I do to the request before it's pushed to traitUpdate($request) the fields are drawn from the setupUpdateOperation() method via getStrippedSaveRequest() -> getAllFieldNames() etc. This means means that I can't actually update a user without updating their password (since it tries to pass NULL for the password and that's no good).

_(Obviously there are other ways to solve this, isolating the password change to it's own method and such, but I want to keep it simple whenever possible here and I think the ability to make a field optional as part of it's definition would be useful in other situations as well)_

**Comment by martijnb92** [comment]
I experienced the same problem, worked around this by inserting a "hidden" field in the crud.

**Comment by drionix** [comment]
> I experienced the same problem, worked around this by inserting a "hidden" field in the crud.

There's a security concern with this workaround because user could just "inspect element" and edit the field. Actually, this is my current workaround for the 'author_id' problem, with extra validation code to make sure that the 'author_id' field has not been altered. And I don't think this is efficient.

Moreover, there's no workaround for the `page_or_link` field problem.

IMHO, the main purpose of `getStrippedSaveRequest()` has been fulfilled by $fillable in laravel model. Therefore, there is no reason to go extra miles and cause the inconvenience.

**Comment by tabacitu** [comment]
Thanks for raising/responding to this issue - I knew it would be a controversial change, and I'm open to going back to the way things were, if it proves that was the better way. 

That being said, let's see if we manage to find decent workarounds for each of your problems, because I do think it's _safer_ to have the request stripped (for most developers and use cases), and I like the fact that you don't _have to_ use ```$fillable``` any more, you can just use ```$guarded``` and you'll still be safe.

### (1) page_or_link not working

@drionix I've just pushed a fix for the ```page_or_link``` field type - it's working now. After a ```composer update``` you should get Backpack 4.0.6 and MenuCrud 2.0.2, where it'll be working. 

**Pro tip: We can now pass an array as a field name, instead of a string.** That's [what I did just now](https://github.com/Laravel-Backpack/MenuCRUD/commit/8b4e9b9d4780d72853caf80b9256dd4da6a14537#diff-4c0570cf6e95730105a00ba8cf5abd7bL52), to fix the page_or_link field type - the "link" and "page_id" attributes did not get saved because they were being stripped out. **It's still possible to have fields that contain multiple inputs, with different names, that all get stored in the database. You just have to mention those attributes inside $field['name'], as an array:**

```diff
            $this->crud->addField([
-                'name' => 'type',
+                'name' => ['type', 'link', 'page_id'],
                'label' => 'Type',
                'type' => 'page_or_link',
                'page_model' => '\Backpack\PageManager\app\Models\Page',
            ]);
```

### (2) adding invisible request data

@drionix , I agree with you here. Adding a hidden input is not always ok because the user might choose to inspect&edit it. @martijnb92 you might want to re-think that strategy, depending on your use case, and who you know uses your forms (trustworthy admins or not). And indeed adding both the hidden input and a validation it for sounds inconvenient, since that's in different files. 

However, we could do:

```php
$this->crud->request->request->add(['author_id'=> backpack_user()->id]);
$this->crud->addField(['type' => 'hidden', 'name' => 'author_id']);
```

This would both add a hidden field (so that the attribute will get saved to the db), and also set the ```author_id``` manually, instead of taking what the user wants.

We could make this process prettier by putting this inside a method, something like:
```php
public function forceAddField($attribute, $value) {
    $this->request->request->add([$attribute => $value]);
    $this->addField(['type' => 'hidden', 'name' => $attribute]);
}
```
Possible names:
- ```forceAddField()```
- ```addHiddenField()``` (confusing imho)
- ```addHiddenFieldWithValue()```
Neither sounds particularly good to me. Thoughts anyone?

### (3) making some fields optional by removing them from the request

@evanrthompson I think I understand your problem. Can you please try removing that _field_ too, in addition to removing it from the request? So you could do:
```php
$this->crud->request->request->remove('password_confirmation');
$this->crud->removeField('password_confirmation');
```
Theoretically, you should be able to only remove the field - no field, no save. Let me know if that works for you.

**Comment by drionix** [comment]
@tabacitu With regard to the workaround for adding/removing request data, I think that's a great workaround.

However, I'm not sure that this workaround will cover every customization other people make with their request data. I mean, some people needed to add a field, some people needed to remove a field. Who knows what other customization people will need to make with their request data?

But for the moment, if adding/removing request data is the only issue with the new `getStrippedSaveRequest()` change, I think adding those workarounds as "customization feature" to the docs in [this](https://backpackforlaravel.com/docs/4.0/crud-operation-create#callbacks) section would be enough.

And if you want to add a method for the workaround, how about `addFieldToRequest()`? Or `addRequestData()`?

**Comment by evanrthompson** [comment]
You know, I think just being able to disable this on a per-model or per-crudController basis would be ideal. I've always really liked $fillable as a methodology but I do understand how it's not universally loved. 

Perhaps in the main `setup()` method just;

`$this->crud->allowArbitraryFields()` or `$this->crud->disableStrictFieldChecking()`
(or something less dumbly named, that is the hardest part :) )

Or failing that if I could just turn this feature off entirely in a config, which is kinda my actual preference here, I like in theory what getStrippedSaveRequest() does but for my purposes it's only an impediment once I've already defined the standard $fillable.

**Comment by pxpm** [comment]
I think the addRequestData() should be enough. In the strip method we can strip everything and manually add the fields that are added using the addRequestData.

**Comment by tabacitu** [comment]
Hey,

I've added examples to add&remove request data to the [Create and Update docs](https://backpackforlaravel.com/docs/4.0/crud-operation-update#callbacks) as @drionix suggested. I think it'll help. @pxpm with the examples there (both to add and remove something from the request), do you still feel like a new method like ```addRequestData()``` is necessary? It would be one line instead of two, yes, but it would also abstract away a manipulation that is fairly easy to do & understand...

@evanrthompson was kind enough to provide a proof-of-concept for making it configurable, which helped create PR https://github.com/Laravel-Backpack/CRUD/pull/2174 - please take a look, I think it fixes both problem (1) and problem (2) above, by allowing devs to choose between ONLY and EXCEPT for all cruds, per crud, or per operation. It's ready to merge from my point of view, but if someone could take a look that would be great - maybe you see something I didn't, or think of a better name;

Thank you for collaborating on this one.
Cheers!

**Comment by guizoxxv** [comment]
I struggled some time on how to add new itens to the request before storing. The documentation in https://backpackforlaravel.com/docs/4.0/crud-operation-create#callbacks was not enought to make it work. Finally I found references on using `$this->crud->setOperationSetting`.

I managed to make it work by adding
`$this->crud->setOperationSetting('saveAllInputsExcept', ['_token', '_method', 'http_referrer', 'current_tab', 'save_action']);`

I believe this section on the documentation should include some reference on this method.

**Comment by guizoxxv** [comment]
I struggled some time on how to add new itens to the request before storing. The documentation in https://backpackforlaravel.com/docs/4.0/crud-operation-create#callbacks was not enought to make it work. Finally I found references on using $this->crud->setOperationSetting.

I managed to make it work by adding
$this->crud->setOperationSetting('saveAllInputsExcept', ['_token', '_method', 'http_referrer', 'current_tab', 'save_action']);

I believe this section on the documentation should include some reference on this method.
